lnst/Tests/PktGen.py

```diff
@@ -0,0 +1,157 @@
+                output = f.readlines()
+            yield device, "\n".join(output)
+
+    def _split_output(self, output: str) -> tuple:
+        match = re.search(r"Params:(.+)Current:(.+)Result:\s(\w+)", output, re.DOTALL)
+        if not match:
+            raise TestModuleError("Could not parse pktgen devide output")
+
+        return match.groups()
+
+    def _parse_values(self, params) -> dict[str, str]:
+        values = {}
+
+        for key, value in re.findall(r"(\w+):?\s(\S+)", params, re.MULTILINE):
+            values[key.lower()] = value
+
+        return values
+
+    def parse_dev_outputs(self):
+        for device, output in self._read_dev_outputs():
+            params, current, state = self._split_output(output)
+
+            self._res[device] = {
+                "params": self._parse_values(params),
+                "current": self._parse_values(current),
+                "state": state,
+            }
+
+
+class PktGen(BaseTestModule):
+    """
+    In the scope of this module, the physical interface is refered as `interface`.
+    Pktgen device (interface@anything) is refered as device.
+
+    Inspired by https://github.com/torvalds/linux/blob/master/samples/pktgen/pktgen_sample03_burst_single_flow.sh
+    """
+
+    cpus = ListParam(type=IntParam())  # each CPU is 1 generator
+
+    pgctrl = StrParam(mandatory=False, default="/proc/net/pktgen/pgctrl")
+
+    src_mac = StrParam(mandatory=False, default="90:e2:ba:ff:ff:ff")
+    dst_mac = StrParam()
+
+    src_ip = IpParam(
+        mandatory=False, default=ipaddress("192.168.1.1")
+    )  # IP doesn't really matter
+    dst_ip = IpParam()
+
+    interface = StrParam()
+
+    count = IntParam(mandatory=False, default=0)  # 0 = no upper limit
+    size = IntParam(mandatory=False, default=60)  # 4 bytes are added for CRC by NIC
+    frags = IntParam(mandatory=False, default=1)
+    burst = IntParam(mandatory=False, default=8)
+
+    duration = IntParam(default=60)
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        if not kmod_in_use("pktgen"):
+            raise TestModuleError("pktgen module is not loaded")
+
+        self._devices = []
+        self._ip6 = True
+        if isinstance(self.params.dst_ip, Ip4Address):
+            self._ip6 = False
+
+        self._res_data = {}
+        self._output_parser = PktGenResultsParser(self._devices, self._res_data)
+
+        self._pg_ctl("reset")
+
+        self._configure_generator()
+
+    def _write_command(self, file: str, cmd: str):
+        with open(file, "w") as f:
+            f.write(cmd + "\n")
+
+    def _pg_ctl(self, cmd: str):
+        self._write_command(self.params.pgctrl, cmd)
+
+    def _pg_thread(self, thread: int, cmd: str):
+        self._write_command(f"/proc/net/pktgen/kpktgend_{thread}", cmd)
+
+    def _pg_set(self, thread: int, cmd: str):
+        self._write_command(f"/proc/net/pktgen/{self.params.interface}@{thread}", cmd)
+
+    def _configure_generator(self):
```

Comment:
can we flip the ordering of these methods? i'd like to see the following order so that it's more natural when reading the class top to bottom:

1. __init__
2. run method
3. configure_generator
4. deconfigure_generator
5. pg_ctl
6. pg thread
7. pg set
8. write command

Reply:
Reordered, thanks.